realm/rfc/0001-backlink-query-syntax.md

```diff
@@ -0,0 +1,232 @@
+  current mixed style found in Realm Java.
+  
+
+# Detailed design
+
+**Example model classes:**
+
+```
+public class Person extends RealmObject {
+	private Dog favoriteDog;
+	private RealmList<Dog> dogs;
+}
+
+public class Dog extends RealmObject {
+	private String name
+
+	// Named backlinks below
+	@Backlink("favoriteDog")
+	private RealmResults<Person> favoriteOwner;
+
+	@Backlink("favoriteDog", "dogs")
+	private RealmResults<Person> allOwners;
+}
+```
+
+**Introduce the following query extension**
+
+- Use `"y[x.z]"` to search field Y on type X where X is a backlink through 
+  field Z.
+- Searching multiple fields should be done in individual query methods:
+  `equalTo("y[x.z]", "foo").equalTo("y[x.z1]", "foo")`
+- This is a supplement to defining the named backlinks.
+- No changes to the migration API.
+- No changes to the schema definition.
+
+Example:
+
+```
+// Named backlink queries
+realm.where(Dog.class).equalTo("allOwners.name", "John").findAll();
+realm.where(Dog.class).equalTo("favoriteOwner.name", "John").findAll();
+
+// Unnamed backlink queries
+realm.where(Dog.class).equalTo("name[Person.favoriteDog]", "John").findAll();
+realm.where(Dog.class)
+  .equalTo("name[Person.favoriteDog]", "John")
+  .equalTo("name[Person.dogs]", "John")
+  .findAll();
+
+// Example in a purely text based variant
+"name[Person.favoriteDog] = 'John'"
+```
+
+**Implementation**
+Implementation is straight forward as we have all schema information available 
+when parsing queries. So simple String manipulation will make it possible to 
+split a field name into a name and a type part.
+
+Validating that the type is an actual backlink with the given field does require 
+an `O(n)` lookup where `m: number of fields in Person`. Since number of fields 
+are generally low, this shouldn't be a problem in practise.
+
+**Advantages**
+- Using "[]" is not an allowed field characters in any of our supported 
+  languages as far as I know. So parsing the above expressions should be 
+  possible with no ambiguity.
+
+- The default case is easy to describe.
+
+- Using [] as a modifier already has precedent in NSPredicate.
+
+
+**Disadvantages**
+
+- Unknown syntax. This concept is not used nor supported in neither NSPredicate, 
+  LINQ or SQL. It will be a "power feature", and might make the language feel 
+  complicated and hard to use.
+
+- Moves the Java query language further away from being type safe.
+
+- Will introduce two ways of querying backlinks.
+
+- Without wildcards for unnamed backlinks queries risk getting extremely verbose.
+
+
+## Alternative syntax suggestions
+
+As an alternative to the above, I have listed a some of variants that were 
+considered:
+
+```
+1) Type in front
+
+- "[Person]name = 'John'"
+- "[Person.favoriteDog]name = 'John'"
+
+
+2) All fields explicit. Group fields using ()
+
+- "name[Person.(favoriteDog, dogs)]  = 'John'"
+
+
+3) Other modifiers (^ up) , (<- back)
+
+- "^Person(favoriteDog, dogs).name"
```

Comment:
`^Person(favoriteDog, dogs).name` for the String based query, this one is my favourite one till now. or maybe use `[]`. Any way `^` makes it clearly a backlink query.